### PR TITLE
CI: add codecov to CI

### PR DESCRIPTION
**Pull request by nawtrey**
* add basic `.codecov.yml` config file to mute
the bot comments

* add code coverage checking to CI matrix

* Fixes issue #421

### CODE REVIEW

**Comment by nawtrey** [comment]
By default the CI should not fail if the coverage drops for a given PR, and the bot comments are turned off. I'm indifferent to which `codecov` features we use, so I just made this fit the requirements @tylerjereddy  requested in #421.

**Comment by tylerjereddy** [comment]
Did you check it on your fork to make sure it works properly? That may be a good first step before asking for permission to activate the hooks here (since you have permission to activate the hooks on your own fork and test there). The codecov settings don't always behave as expected--I know some projects that basically had to "ban" the bot/user.

File: darshan-util/pydarshan/.codecov.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+# default coverage settings
+coverage:
+  precision: 2
+  round: down
+  range: "70...100"
+
+# default parser settings
+parsers:
+  gcov:
```
Comment (tylerjereddy) [inline darshan-util/pydarshan/.codecov.yml:14]: `gcov` is for line coverage of compiled code, but you haven't activated coverage builds of the C source underneath the hood, so I'm not sure I understand the motivation for the Python side of things.

We could discuss that if we want it for the C code eventually, but that's probably another can of worms we don't want to deal with initially.

Also, did you check that `.codecov.yml` is properly picked up when not at the root of the repository proper (it is usually placed there)

File: darshan-util/pydarshan/.codecov.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# Codecov configuration to make it a bit less noisy
+codecov:
+  # don't fail the CI
+  require_ci_to_pass: yes
```
Comment (tylerjereddy) [inline darshan-util/pydarshan/.codecov.yml:4]: can you explain the reasoning for choosing this setting vs: https://github.com/scipy/scipy/blob/master/codecov.yml#L2

Also, maybe discuss some of the reasons for not adopting the battle-tested settings there or in other OSS projects?

**Comment by tylerjereddy** [comment]
Annotations are annoying too--they should be disabled like here: https://github.com/scipy/scipy/blob/master/codecov.yml#L23

Basically, build up from battle-tested examples..

**Comment by tylerjereddy** [comment]
Another comment here is that you're trying to run `codecov` when you don't even have a coverage report file to begin with--the `pytest` command would have to be altered to generate a coverage report in the appropriate format, and you'll need to make sure the coverage-related deps are installed.

**Comment by tylerjereddy** [comment]
any update here?

**Comment by tylerjereddy** [comment]
Nik has draft version here: https://github.com/nawtrey/darshan/pull/1

Looks closer now

File: .github/workflows/linux.yml [hunk 3]
```diff
@@ -44,7 +44,7 @@ jobs:
           # the test suite is sensitive to
           # relative dir for test file paths
           cd darshan-util/pydarshan
-          pytest
+          pytest --cov-report xml --cov=darshan
```
Comment (tylerjereddy) [inline .github/workflows/linux.yml:47]: may want to cover the folder with the test modules in it as well--that can help catch cases with tests that accidentally have the same name for example

Comment (nawtrey) [reply .github/workflows/linux.yml]: I've added `--cov=tests` to the above command. I tested it [on my fork](https://github.com/nawtrey/darshan/pull/2) and `/tests` showed up in the `codecov` GUI

**Comment by tylerjereddy** [comment]
I think the only thing left here is to point to an example on your fork where we can see the "%" change for the "patch" entry like this: 

image

I believe it *should* work if you open a PR on your fork against a branch that has these changes merged in and the "base data" registered with codecov.

**Comment by nawtrey** [comment]
> I think the only thing left here is to point to an example on your fork where we can see the "%" change for the "patch" entry like this:
> 
> image
> 
> I believe it _should_ work if you open a PR on your fork against a branch that has these changes merged in and the "base data" registered with codecov.

I had my forked `pydarshan-devel` branch merged with my original codecov-related [PR](https://github.com/nawtrey/darshan/pull/1), and when I opened the [new PR](https://github.com/nawtrey/darshan/pull/2) it did exhibit that behavior (it gave me the % of diff hit and the target). 

You can view it by clicking the checkmark for the commit just above [this comment](https://github.com/nawtrey/darshan/pull/2#issuecomment-901975301)(the hash in the photo):
Screen Shot 2021-08-20 at 1 15 51 PM

**Comment by tylerjereddy** [comment]
Ok, I'll put this in and comment on the final steps to close the matching issue.

**Pull request merged**